core/src/main/java/com/scalar/db/transaction/jdbc/JdbcTransaction.java

```diff
@@ -163,6 +178,58 @@ public void commit() throws CommitException, UnknownTransactionStatusException {
     }
   }
 
+  private void throwIfMutationHasUnsatisfiedCondition()
+      throws CommitUnsatisfiedConditionException, UnknownTransactionStatusException {
+    if (mutationWithUnsatisfiedCondition == null) {
+      return;
+    }
+
+    try {
+      connection.rollback();
+      throwUnsatisfiedConditionException(mutationWithUnsatisfiedCondition);
+    } catch (SQLException sqlException) {
+      throw new UnknownTransactionStatusException("failed to rollback", sqlException, txId);
+    } finally {
+      try {
+        connection.close();
+      } catch (SQLException e) {
+        logger.warn("failed to close the connection", e);
+      }
+    }
+  }
+
+  private void throwUnsatisfiedConditionException(Mutation mutation)
+      throws CommitUnsatisfiedConditionException {
+    assert mutation.getCondition().isPresent();
+
+    // Build the exception message
+    MutationCondition condition = mutation.getCondition().get();
+    String conditionColumns = "";
```

Comment:
How about using StringBuilder instead to avoid instantiating temporary objects?

Reply:
See also https://dev.to/this-is-learning/performance-benchmarking-string-and-string-builder-3bid

Reply:
That is a good point, thank you. 
see https://github.com/scalar-labs/scalardb/pull/899/commits/494137ed46c6343b4b8d223130dfd5b76a2b8f24